src/metabase/query_processor/pivot.clj

```diff
@@ -1,98 +1,144 @@
 (ns metabase.query-processor.pivot
   "Pivot table actions for the query processor"
   (:require [clojure.core.async :as a]
-            [clojure.set :refer [map-invert]]
+            [metabase.mbql.normalize :as normalize]
             [metabase.query-processor :as qp]
             [metabase.query-processor.context :as qp.context]
-            [metabase.query-processor.store :as qp.store]))
+            [metabase.query-processor.context.default :as context.default]
+            [metabase.query-processor.error-type :as qp.error-type]
+            [metabase.query-processor.store :as qp.store]
+            [metabase.util.i18n :refer [tru]]
+            [schema.core :as s]))
 
-(defn powerset
-  "Generate a powerset while maintaining the original ordering as much as possible"
-  [xs]
-  (for [combo (reverse (range (int (Math/pow 2 (count xs)))))]
-    (for [item  (range 0 (count xs))
-          :when (not (zero? (bit-and (bit-shift-left 1 item) combo)))]
-      (nth xs item))))
+(defn- group-number
+  "Come up with a display name given a combination of breakout `indecies` e.g.
 
-(defn- generate-specified-breakouts
-  "Generate breakouts specified by pivot_rows"
-  [breakouts pivot-rows pivot-cols]
-  (let [generator (fn [coll]
-                    (for [pivots coll]
-                      (map #(nth breakouts %) pivots)))]
-    (concat
-     [breakouts]
-     (generator [pivot-cols])
-     (when (and (seq pivot-cols)
-                (seq pivot-rows))
-       (generator [(concat [(first pivot-rows)] pivot-cols)]))
-     (generator (powerset (or pivot-rows
-                              ;; this can happen for the public/embed endpoints,
-                              ;; where we aren't given a pivot_rows / pivot_cols
-                              ;; parameter, so we'll just generate everything
```

Comment:
I forgot to mention in the issue what this comment calls out. We need to handle missing `pivot-rows` and `pivot-cols`.

The frontend doesn't have the query when public/embedded so it can't generate the list of indexes. Two possible hacks around that:
1. When those are missing/empty return a full powerset.
2. Duplicate the logic to generate rows/cols indices from query+vis settings in the backend.

This issue should go away when we build `breakout_sets` into MBQL. Then, the query itself will contain the needed information.

For now, either option works for me, but I prefer 1. It's slower for public/embedded since it runs unnecessary queries, but it avoids potentially inconsistent FE/BE logic duplication.

Reply:
So... use the full powerset if cols/rows are missing. Otherwise use the rules from your issue

Reply:
Yup that's right!